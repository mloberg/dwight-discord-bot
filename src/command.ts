import { SlashCommandBuilder, SlashCommandIntegerOption } from '@discordjs/builders';
import { CommandInteraction } from 'discord.js';

type Handler = (command: CommandInteraction) => Promise<void>;

export class CommandBuilder extends SlashCommandBuilder {
    constructor(readonly handler: Handler) {
        super();
    }
}

export const rollOption = (dice: string): SlashCommandIntegerOption =>
    new SlashCommandIntegerOption().setName('roll').setDescription(`Value of a ${dice} roll.`);
