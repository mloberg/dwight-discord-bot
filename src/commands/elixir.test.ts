import { Client, Guild, Message, TextChannel } from 'discord.js';

import command, { elixirs } from './elixir';

const mocks = {
    delete: jest.fn(),
    reply: jest.fn(),
};

jest.mock('discord.js', () => ({
    Client: jest.fn(),
    Guild: jest.fn(),
    TextChannel: jest.fn(),
    Collection: jest.fn(),
    Message: jest.fn().mockImplementation(() => {
        return {
            delete: mocks.delete,
            reply: mocks.reply,
        };
    }),
}));

describe('_elixir configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('elixir');
        expect(command.description).toEqual('Craft an experimental elixir');
    });

    it('should have aliases', () => {
        expect(command.alias).toContain('elx');
    });
});

describe('_elixir', () => {
    let message: Message;

    beforeEach(() => {
        mocks.delete.mockClear();
        mocks.reply.mockClear();
        mocks.reply.mockReturnThis();

        const client = new Client();
        const guild = new Guild(client, {});
        const channel = new TextChannel(guild, {});
        message = new Message(client, {}, channel);
    });

    it('returns an elixir', async () => {
        const reply = await command.run(message, { $0: 'elixir', _: [] });

        expect(mocks.delete).toHaveBeenCalledTimes(1);
        expect(reply).toBe(message);

        const elixir = mocks.reply.mock.calls[0][0];
        expect(elixirs).toContainEqual(elixir);
    });

    it('returns an elixir for a dice roll', async () => {
        const one = await command.run(message, { $0: 'elixir', _: ['1'] });
        const two = await command.run(message, { $0: 'elixir', _: [], roll: 4 });

        expect(mocks.delete).toHaveBeenCalledTimes(2);
        expect(one).toBe(message);
        expect(two).toBe(message);

        expect(mocks.reply).toHaveBeenCalledWith(
            '**Healing**. The drinker regains a number of hit points equal to 2d4 + your Intelligence modifier.',
        );
        expect(mocks.reply).toHaveBeenCalledWith(
            '**Boldness**. The drinker can roll a d4 and add the number rolled to every attack roll and saving throw they make for the next minute.',
        );
    });
});
