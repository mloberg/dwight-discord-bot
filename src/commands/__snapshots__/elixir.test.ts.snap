// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/elixir is a slash command 1`] = `
Object {
  "config": Object {
    "default_permission": undefined,
    "description": "Craft an experimental elixir",
    "name": "elixir",
    "options": Array [
      Object {
        "choices": undefined,
        "description": "Value of a d6 roll.",
        "name": "roll",
        "required": false,
        "type": 4,
      },
    ],
  },
  "handle": [Function],
}
`;
